src/internal/connector/graph/retry_middleware.go

```diff
@@ -0,0 +1,131 @@
+package graph
+
+import (
+	"context"
+	"io"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	abs "github.com/microsoft/kiota-abstractions-go"
+	khttp "github.com/microsoft/kiota-http-go"
+)
+
+// ---------------------------------------------------------------------------
+// Client Middleware
+// ---------------------------------------------------------------------------
+
+// RetryHandler handles transient HTTP responses and retries the request given the retry options
+type RetryHandler struct {
+	// default options to use when evaluating the response
+	options RetryHandlerOptions
+}
+
+type RetryHandlerOptions struct {
+	// request never retried if flag set to true
+	NoRetry bool
+	// The maximum number of times a request can be retried
+	MaxRetries int
```

Comment:
scenario A-
- user wants No Retries-
```
input provided-
MaxRetries- 0
```

- user wants default Retries-

```
no input provided. But since `MaxRetries` is a `int` we get the default value 0-
MaxRetries -0
```
In both the scenarios we will get 0 as `MaxRetries` and it would be difficult to decide retries.
Please correct me if I am missing something in this scenario.

Reply:
The current code is written in a way that - if `noRetries` is set to true no other configuration will be taken in consideration.

Reply:
Since default value of `MaxRetries(type int)` is 0. It was difficult to know if 0 was send or input or if it is a default value. 
To handle this, 
- we ask user for graph.MaxRetries(). If provided as graph.MaxRetries(0), we internally set `noRetry` flag to true. 
Please suggest if there is a better way to handle it.

Reply:
I'd recommend adding the flag to the constructor options, but not the retryHandler struct.  Ex:
```go
type RetryHandler struct {
    MaxRetries int
    MinDelay time.Duration
}

type clientConfig struct {
    noTimeout bool
    maxRetires int
    overrideRetryCount bool
}

func MaxRetries(max int) option {
    return func(c *clientConfig) {
        c.overrideRetryCount = true
        c.maxRetries = max
    }
}

func (c *clientConfig) applyMiddlewareConfig(handler *RetryHandler) {
    if c.overrideRetryCount {
         handler.maxRetries = c.maxRetries
    }
}
```